geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/FeatureSpecificReader.scala

```diff
@@ -24,28 +24,39 @@ import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 
 import scala.collection.JavaConversions._
 
-class FeatureSpecificReader(oldType: SimpleFeatureType, newType: SimpleFeatureType,
-                            opts: Set[SerializationOption] = Set.empty)
+class FeatureSpecificReader(var oldType: SimpleFeatureType, var newType: SimpleFeatureType,
+                            opts: SerializationOptions = SerializationOptions.none)
   extends DatumReader[AvroSimpleFeature] {
 
   def this(sft: SimpleFeatureType) = this(sft, sft)
 
-  var oldSchema = generateSchema(oldType)
-  val fieldsDesired = DataUtilities.attributeNames(newType).map(encodeAttributeName)
+  // DataFileStreams or DataFileReaders may set this after construction of the object
+  // so all instance variables need to be be lazy
+  var oldSchema = Option(oldType).map(generateSchema(_, opts.withUserData)).orNull
 
-  def isDataField(f: Schema.Field) =
-    !f.name.equals(FEATURE_ID_AVRO_FIELD_NAME) && !f.name.equals(AVRO_SIMPLE_FEATURE_VERSION)
-
-  override def setSchema(schema: Schema): Unit = oldSchema = schema
+  lazy val fieldsDesired = DataUtilities.attributeNames(newType).map(encodeAttributeName)
 
-  val dataFields = oldSchema.getFields.filter { isDataField }
+  lazy val dataFields = oldSchema.getFields.filter { isDataField }
 
-  val typeMap: Map[String, Class[_]] =
+  lazy val typeMap: Map[String, Class[_]] =
     oldType.getAttributeDescriptors.map { ad => encodeAttributeName(ad.getLocalName) -> ad.getType.getBinding }.toMap
 
-  val nillableAttrs: Set[String] = oldType.getAttributeDescriptors.filter(_.isNillable).map {
-      ad => encodeAttributeName(ad.getLocalName)
-    }.toSet
+  lazy val nillableAttrs: Set[String] = oldType.getAttributeDescriptors.filter(_.isNillable).map {
+    ad => encodeAttributeName(ad.getLocalName)
+  }.toSet
+
+  def isDataField(f: Schema.Field) =
+    !f.name.equals(FEATURE_ID_AVRO_FIELD_NAME) &&
+      !f.name.equals(AVRO_SIMPLE_FEATURE_VERSION) &&
+      !f.name.equals(AVRO_SIMPLE_FEATURE_USERDATA)
+
+  override def setSchema(schema: Schema): Unit = oldSchema = schema
+
+  def setTypes(oldType: SimpleFeatureType, newType: SimpleFeatureType): Unit = {
```

Comment:
Is this sort of mutability a requirement of Avro, or is there expensive internal state that we are somehow conserving?